test/interface/test_sorting.py

```diff
@@ -109,43 +105,67 @@ def test_invalid_direction(self, session):
 
 class TestSortApplied(object):
 
-    @pytest.mark.usefixtures('multiple_bars_inserted')
+    """Tests that results are sorted only according to the provided
+    filters.
+
+    Does NOT test how rows with the same values are sorted since this is
+    not consistent across RDBMS.
+
+    Does NOT test whether `NULL` field values are placed first or last
+    when sorting since this may differ across RDBMSs.
+
+    SQL defines that `NULL` values should be placed together when
+    sorting, but it does not specify whether they should be placed first
+    or last.
+    """
+
+    @pytest.mark.usefixtures('multiple_bars_with_no_nulls_inserted')
     def test_single_sort_field_asc(self, session):
         query = session.query(Bar)
         order_by = [{'field': 'name', 'direction': 'asc'}]
 
         sorted_query = apply_sort(query, order_by)
         result = sorted_query.all()
 
+        results_with_same_order_1 = [
+            result[0].id, result[1].id, result[2].id, result[3].id
+        ]
+        results_with_same_order_2 = [result[5].id, result[6].id]
```

Comment:
I still find this confusing, but now recognise that it is my fault for setting up the fixtures the way I did. Actually `result_same_name_1` and `result_same_name_4` were quite appropriate names!

Thinking about it a bit more though, why is the test making assertions about the IDs? Wouldn't it be sufficient to just assert that the _names_ of the results are in ascending order? i.e.

``` python
results = sorted_query.all()
assert [result.name for result in results] == ["name_1", "name_1", ..., "name_5"]
```

Reply:
(Recognise this is my fault too -- I wrote the original test!)

Reply:
Yes, that looks better! I've changed the tests as you suggested.

I have also s/result/results in the rest of the tests.

However, there are some other tests where we sort by multiple fields and it gets a bit more complicated to use this way of testing (as the number of filters grows). It seems easier to keep the assertions based on the IDs (since we just test the order of the rows), instead of adding all the field values to all the tests. On the other hand, by doing so you can also explicitely see how the fields are ordered:

```python
assert len(results) == 8
assert results[0].id == 1
assert results[1].id == 3
assert results[2].id == 7
assert results[3].id == 5
assert results[4].id == 2
assert results[5].id == 6
assert results[6].id == 4
assert results[7].id == 8
```

vs.

```python
assert [
    (result.name, result.count, result.id) for result in results
] == [
    ('name_1', 5, 1),
    ('name_1', 3, 3),
    ('name_1', 2, 7),
    ('name_1', 2, 5),
    ('name_2', 10, 2),
    ('name_4', 15, 6),
    ('name_4', 12, 4),
    ('name_5', 1, 8),
]
```

What do you think @mattbennett ? Would you apply the latter way of testing it to all the tests in that module?

Reply:
I like the second way! All the tests consistently using this would be nice. Adding a simple comment saying "expect results to be ordered by name, then count" (or whatever) would make things very clear.

Reply:
I've amended the assertions as you suggested @mattbennett and it looks way more readable!

It's easy now to compare the results with the actual filters placed a few lines above the assertions and visually check that it's sorted correctly. I think extra comments are not even needed at this point, having all the info that we need in the list comprehension (and the `order_by` dicts).

Please take another look when you have a moment.

Reply:
Much nicer!